cloudstack_plugin/vpn.py

```diff
@@ -0,0 +1,446 @@
+    """
+    Returns the VPN Gateway for a VPC.
+    """
+    for obj in cloud_driver.ex_list_vpn_gateways():
+        if vpc_id == obj.vpc_id:
+            return obj
+
+
+def get_vpn_customer_gateway(cloud_driver, id_):
+    """
+    Returns the VPN Customer Gateway.
+    """
+    vpn_customer_gateways = cloud_driver.ex_list_vpn_customer_gateways()
+
+    if not len(vpn_customer_gateways):
+        return
+
+    for obj in vpn_customer_gateways:
+        if id_ == obj.id:
+            return obj
+
+
+def get_vpn_connections(cloud_driver):
+    """
+    Returns VPN Connections.
+    """
+    return cloud_driver.ex_list_vpn_connections()
+
+
+def get_vpn_connection(cloud_driver, id_):
+    """
+    Returns the VPN Connection.
+    """
+    vpn_connections = get_vpn_connections(cloud_driver)
+
+    if not len(vpn_connections):
+        return
+
+    for obj in vpn_connections:
+        if id_ == obj.id:
+            return obj
+
+
+def get_vpn_connection_by_gateways(cloud_driver, vpn_gateway_id,
+                                   vpn_customer_gateway_id):
+    vpn_connections = cloud_driver.ex_list_vpn_connections()
+
+    if not len(vpn_connections):
+        return
+
+    for obj in vpn_connections:
+        if ((obj.vpn_customer_gateway_id == vpn_customer_gateway_id) and
+                (obj.vpn_gateway_id == vpn_gateway_id)):
+            return obj
+
+
+def vpn_gateway_exists(cloud_driver, id_):
+    """
+    Returns if the VPN Gateway exists.
+    """
+    if get_vpn_gateway(cloud_driver, id_):
+        return True
+
+    return False
+
+
+def vpn_gateway_exists_for_vpc(cloud_driver, vpc_id):
+    """
+    Returns if the VPN Gateway exists for the VPC.
+    """
+    vpn_gateway = get_vpn_gateway_by_vpc(cloud_driver, vpc_id)
+
+    if vpn_gateway:
+        return True
+
+    return False
+
+
+def vpn_customer_gateway_exists(cloud_driver, id_):
+    """
+    Returns if the VPN Customer Gateway exists.
+    """
+    if get_vpn_customer_gateway(cloud_driver, id_):
+        return True
+
+    return False
+
+
+def vpn_connection_exists(cloud_driver, id_):
+    """
+    Returns if the VPN Connection exists.
+    """
+    if get_vpn_connection(cloud_driver, id_):
+        return True
+
+    return False
+
+
+def get_vpc_routers(cloud_driver, vpc_id):
+    """
+    Returns the list of VPC Router objects.
+    """
+    vpc_routers = cloud_driver.ex_list_routers(vpc_id=vpc_id)
+
+    if vpc_routers:
+        return vpc_routers
+
+
+def wait_for_vpc_routers_to_stabilize(ctx, cloud_driver, vpc_id):
+    while True:
+        vpc_routers = get_vpc_routers(cloud_driver, vpc_id)
+
+        if not vpc_routers:
+            return
+
+        is_running = all(
+            [obj.state == VPC_ROUTER_STATE_RUNNING for obj in vpc_routers])
+
+        if is_running:
+            ctx.logger.info('VPC Routers running, continue...')
+            return
+        else:
+            ctx.logger.info('VPC Routers not running, wait...')
```

Comment:
When a plugin require to do polling for a certain state, instead of while True,  the best practice is to use ctx.operation.retry in order to free the worker for executing other pending tasks. More info can be found here: http://getcloudify.org/guide/3.2/guide-plugin-creation.html#asynchronous-operations